### ISSUE TITLE
Unexpected keyword argument "retain_graph" for "tensor.backward"

### ISSUE THREAD

**Issue by DavidS3141** [pytorch/pytorch#37818]
## 🐛 Bug

`mypy` throws an error on `some_tensor.backward(retain_graph=True)` complaining that `retain_graph` is not an acceptable keyword argument. When `reveal_type`ing the backward function, it gets clear that `mypy` associates the `torch.backward` function. Now this is just a guess that this might need fixing in pytorch but it could also be a bug on `mypy`'s side but I figured I have to litte information for them so I started here.

## To Reproduce

1. Well, really just create a pytorch tensor and call `.backward(retain_graph)` and let `mypy` run over this.

For me it looks like this:
![image](https://user-images.githubusercontent.com/6333870/81052335-8804bc00-8ec3-11ea-94d5-f08982aebd43.png)
![image](https://user-images.githubusercontent.com/6333870/81052351-918e2400-8ec3-11ea-87f6-2028f7695a31.png)

## Expected behavior

I expect it to show no type error as retain_graph is a well-documented keyword argument of backward.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5.0+cu92
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip3
 - Build command you used (if compiling from source):
 - Python version: 3.6.9
 - CUDA/cuDNN version: 10.0

cc @ezyang

**Comment by ezyang** [comment]
Thank you for the report. If someone can figure out what's going on here we'd very happily accept a patch to fix this.

**Comment by rgommers** [comment]
@DavidS3141 `torch.backward` does not exist, and `torch.autograd.backward` does have a `retain_graph` keyword. This typechecks fine:
```
import torch

x = torch.ones(2, 2, requires_grad=True)
y = x + 1
z = y * y * 3
out = z.mean()
out.backward(retain_graph=True)
```

Could you please give a reproducible example?

**Comment by DavidS3141** [comment]
That is interesting, for me your example does not work, e.g. it produces exactly the mypy typecheck error I was reporting (see below).

I am a bit stumped about `torch.autograd.backward` having the keyword, I thought I saw a function declaration somewhere without it. Might this be a version issue, because I use pytorch 1.5.0 for CUDA 9.2 instead of 1.5.0 for CUDA 10.1 (or whatever is the most recent).

For completeness, my `mypy` version is 0.770.
```python
$ mypy --version
mypy 0.770

$ cat test_from_issue_pytorch_37818.py 
#!/usr/bin/env python3

# this relates to the pytorch issue 37818
# https://github.com/pytorch/pytorch/issues/37818

import torch

x = torch.ones(2, 2, requires_grad=True)
y = x + 1
z = y * y * 3
out = z.mean()
reveal_type(out.backward)
out.backward(retain_graph=True)

$ mypy test_from_issue_pytorch_37818.py 
test_from_issue_pytorch_37818.py:12: note: Revealed type is 'def (gradient: Union[torch.Tensor, None] =, keep_graph: builtins.bool =, create_graph: builtins.bool =)'
test_from_issue_pytorch_37818.py:13: error: Unexpected keyword argument "retain_graph" for "backward" of "Tensor"
Found 1 error in 1 file (checked 1 source file)
```

**Comment by rgommers** [comment]
Thanks, that helps. In a clean conda env with pytorch 1.5.0 and mypy 0.770 I can reproduce your issue. In master it is fixed:
```
$ mypy --check-untyped-defs torch_37818.py 
torch_37818.py:8: note: Revealed type is 'def (gradient: Union[torch.Tensor, None] =, keep_graph: builtins.bool =, create_graph: builtins.bool =)'
torch_37818.py:9: error: Unexpected keyword argument "retain_graph" for "backward" of "Tensor"
Found 1 error in 1 file (checked 1 source file)

(pytorch15) rgommers@machine:~/code/tmp$ conda activate pytorch
(pytorch) rgommers@machine:~/code/tmp$ mypy --check-untyped-defs torch_37818.py 
torch_37818.py:8: note: Revealed type is 'def (gradient: Union[torch.Tensor, None] =, retain_graph: Union[builtins.bool, None] =, create_graph: builtins.bool =)'
```

Type annotations are improving quickly at the moment; the fix for this will be in 1.6.0. Thanks for the report @DavidS3141.
